Improvements to hdmi-codec and conversion of vc4-hdmi

mripard opened the pull request:
Hi,

Here's a rework of the HBR passthrough and channel mapping support for vc4-hdmi to rely on the hdmi-codec driver instead of duplicating some of its code.

I've checked that the reported ELD is still valid (using amixer -c0 cget numid=$ID), and iecset to dump the iec status.

I don't have the hardware to test this (I think?) so it would be great if someone could test and make sure nothing got broken in the process. I'm especially worried about the ordering of the calls that is mentioned here: https://github.com/raspberrypi/linux/pull/3717/commits/06a9ff06ad3435e67a8377bdc2e2985d6764c737

I couldn't find in Kodi, LibreElec or alsa-lib where the status setup was done and the order relative to the hw_params call so it might be broken at the moment.

Thanks!
Maxime

siajeeheng committed:
ASoC: codec: hdmi-codec: Support IEC958 encoded PCM format

Existing hdmi-codec driver only support standard pcm format.
Support of IEC958 encoded format pass from ALSA IEC958 plugin is needed
so that the IEC958 encoded data can be streamed to the HDMI chip.

Signed-off-by: Sia Jee Heng <jee.heng.sia@intel.com>
Link: https://lore.kernel.org/r/20210204014258.10197-2-jee.heng.sia@intel.com
Signed-off-by: Mark Brown <broonie@kernel.org>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: Register HDMI codec

The hdmi-codec brings a lot of advanced features, including the HDMI
channel mapping. Let's use it in our driver instead of our own codec.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
drm/vc4: hdmi: Remove redundant variables

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
ASoC: hdmi-codec: Rework to support more controls

We're going to add more controls to support the IEC958 output, so let's
rework the control registration a bit to support more of them.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
snd: iec958: split status creation and fill

In some situations, like a codec probe, we need to provide an IEC status
default but don't have access to the sampling rate and width yet since
no stream has been configured yet.

Each and every driver has its own default, whereas the core iec958 code
also has some buried in the snd_pcm_create_iec958_consumer functions.

Let's split these functions in two to provide a default that doesn't
rely on the sampling rate and width, and another function to fill them
when available.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

mripard committed:
ASoC: hdmi-codec: Add iec958 controls

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix committed:
ARM: dts: bcm2711: Tune DMA parameters for HDMI audio

Enable NO_WAIT_RESP, DMA_WIDE_SOURCE, DMA_WIDE_DEST, and bump the DMA
panic and AXI priorities to avoid any DMA transfer error with HBR audio
(8 channel, 192Hz).

Signed-off-by: Dom Cobley <popcornmix@gmail.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>

popcornmix commented:
I've just given this a test and TrueHD passthrough worked but DTS-HD (despite making DTS-HD led on AVR come on) was silent.
Possibly related to the iec_status bits, but @HiassofT knows more about this.

popcornmix commented:
I'm wondering about the AES bits set here: https://github.com/xbmc/xbmc/blob/0e7f7628cb5fe8453fedf801b12cc659d2eb7bcb/xbmc/cores/AudioEngine/Sinks/AESinkALSA.cpp#L453
I think they get passed to driver and affect whether we are in PCM or passthrough mode.

HiassofT commented:
We can't select between PCM and HBR frames in hw_params, as the alsa hook plugin passes on the IEC status bits after hw_params (and before prepare) - therefore I moved the code to prepare.

See also the commit message of my change https://github.com/raspberrypi/linux/commit/fd45fc3c28ef696d5d55861f5d98ee713a7e2dfb

jernejsk commented on sound/soc/codecs/hdmi-codec.c:
```diff
@@ -278,6 +278,7 @@ struct hdmi_codec_priv {
 	bool busy;
 	struct snd_soc_jack *jack;
 	unsigned int jack_status;
+	u8 iec_status[4];
```
Maybe it would be better to have 5 byte array? Although fifth byte doesn't seem to be used usually, it's nice to have this option in shared codec (DW-HDMI driver uses that fifth byte).

mripard replied:
Indeed, I changed it, thanks!

mripard committed:
ASoC: hdmi-codec: Add a prepare hook

The IEC958 status bit is usually set by the userspace after hw_params
has been called, so in order to use whatever is set by the userspace, we
need to implement the prepare hook. Let's add it to the hdmi_codec_ops,
and mandate that either prepare or hw_params is implemented.

Signed-off-by: Maxime Ripard <maxime@cerno.tech>

jernejsk commented on include/sound/hdmi-codec.h:
```diff
@@ -34,11 +34,6 @@ struct hdmi_codec_daifmt {
-	 * IEC958_SUBFRAME format to the underneath driver.
-	 */
-	snd_pcm_format_t bit_fmt;
 };
```
Please don't do that, I'm using it in currently out of tree changes for DW-HDMI audio, so I was pretty happy to see it's added.

HiassofT commented on sound/soc/codecs/hdmi-codec.c:
```diff
@@ -619,6 +662,20 @@ static const struct snd_soc_dai_ops hdmi_codec_spdif_dai_ops = {
 struct snd_kcontrol_new hdmi_codec_controls[] = {
+	{
+		.access = SNDRV_CTL_ELEM_ACCESS_READ,
+		.iface = SNDRV_CTL_ELEM_IFACE_MIXER,
```
this should be SNDRV_CTL_ELEM_IFACE_PCM like in the IEC958 Playback Default and ELD controls below

mripard replied:
Hi,
The doc seems to suggest otherwise though, from https://www.kernel.org/doc/html/latest/sound/kernel-api/writing-an-alsa-driver.html#iec958-s-pdif

> “IEC958 Playback Con Mask” is used to return the bit-mask for the IEC958 status bits of consumer mode. Similarly, “IEC958 Playback Pro Mask” returns the bitmask for professional mode. They are read-only controls, and are defined as MIXER controls (iface = SNDRV_CTL_ELEM_IFACE_MIXER).

HiassofT replied:
Having mask and default controls on different interfaces would be really odd as these are closely related to each other.

Current kernel code shows mask/default controls are either all defined as mixer controls (eg hda) or pcm controls (eg ice1712 and intel hdmi).

see eg `git grep -B 2 "SNDRV_CTL_NAME_IEC958.*PLAYBACK" sound`

mripard replied:
Some have the doc suggest too (ice1724, oxygen, trident). It's a bit of a mess, I'll try to clear it up while upstreaming :)

mripard commented:
I just pushed a new version addressing the comments from @jernejsk  and @HiassofT

popcornmix commented:
Just tested update PR and TrueHD passthrough is still silent for me. Removing this PR and audio is back.

HiassofT commented:
@popcornmix did you update vc4-hdmi.conf? https://github.com/HiassofT/LibreELEC.tv/commit/bbde777b672fdf84dc6317fc0df2c10f781b7173

popcornmix commented:
No - let me try that.

popcornmix commented:
Yes, that seems to be working for a range of passthrough formats and multichannel PCM.
Looks like a nice clean up.

@HiassofT I assume it was fine for you?

HiassofT commented:
Yes, it's fine for me.

Haven't tested the latest version (and can't test multichannel PCM output / channel mapping) but I glanced through the changes and they looked OK to me.

Might be worth sending out an RFC series and see what upstream ASoC/dri folks think about the hdmi-codec changes

mripard commented:
It's prepared already, I was waiting for your testing results to send it. If it works as expected for you, I'll send it in the afternoon

mripard commented:
I just did:
https://lore.kernel.org/dri-devel/20210507140334.204865-1-maxime@cerno.tech/T/#mab1b0e61ab14460321e455b1f28552a88ded3e5d

mripard commented:
The ASoC parts have been accepted upstream and are on their way to 5.13

popcornmix commented:
Does this PR still match upstream version? (not sure if it's been reworked since this PR's version).

mripard commented:
The commits are not identical in the ASoC part but it's only about the commit logs and some additional documentation I wrote in order to upstream. The code itself is identical.

The DRM side is fairly different but it's mostly due to fact that we had to revert the old code in the downstream tree, while we only had to add new code upstream.

popcornmix commented:
@pelwell I think this should be merged. I've been running with it for a while at home without issue and it's merged upstream.

pelwell commented:
Some of these don't apply cleanly to rpi-5.12.y (I've merged the ones that do).
